packages/lesswrong/components/comments/CommentsTableOfContents.tsx

```diff
@@ -0,0 +1,189 @@
+import React from 'react';
+import { Components, registerComponent } from "../../lib/vulcan-lib";
+import { CommentTreeNode } from '../../lib/utils/unflatten';
+import { getCurrentSectionMark, getLandmarkY, ScrollHighlightLandmark, useScrollHighlight } from '../hooks/useScrollHighlight';
+import { useLocation, useNavigation } from '../../lib/routeUtil';
+import isEmpty from 'lodash/isEmpty';
+import qs from 'qs'
+import { userGetDisplayName } from '../../lib/collections/users/helpers';
+import { commentsTableOfContentsEnabled } from '../../lib/betas';
+
+const styles = (theme: ThemeType): JssStyles => ({
+  root: {
+    color: theme.palette.text.dim,
+  },
+  comment: {
+    display: "inline-flex",
+  },
+  commentKarma: {
+    width: 20,
+  },
+  commentAuthor: {
+  },
+  collapseButtonWrapper: {
+    marginLeft: 4,
+    height: 24,
+  },
+  collapseButton: {
+    "&:hover": {
+      background: theme.palette.greyAlpha(0.1),
+      borderRadius: 8,
+    },
+  },
+  collapseButtonExpanded: {
+  },
+  collapseButtonCollapsed: {
+    transform: "rotate(-90deg)",
+  },
+  postTitle: {
+    minHeight: 24,
+    paddingTop: 4,
+  },
+})
+
+const CommentsTableOfContents = ({commentTree, answersTree, post, classes}: {
+  commentTree?: CommentTreeNode<CommentsList>[],
+  answersTree?: CommentTreeNode<CommentsList>[],
+  post: PostsWithNavigation | PostsWithNavigationAndRevision,
+  classes: ClassesType,
+}) => {
+  const { TableOfContentsRow } = Components;
+  const flattenedComments = flattenCommentTree([
+    ...(answersTree ?? []),
+    ...(commentTree ?? [])
+  ]);
+  const { landmarkName: highlightedLandmarkName } = useScrollHighlight(
+    flattenedComments.map(comment => commentIdToLandmark(comment._id))
+  );
+  
+  if (!commentsTableOfContentsEnabled) {
+    return null;
+  }
+  
+  return <div className={classes.root}>
+    <TableOfContentsRow key="postTitle"
+      href="#"
+      onClick={ev => {
+        window.scrollTo({
+          top: 0,
+          behavior: "smooth"
+        });
+      }}
+      highlighted={highlightedLandmarkName==="above"}
+      title
+    >
+      <span className={classes.postTitle}>
+        {post.title?.trim()}
+      </span>
+    </TableOfContentsRow>
+
+    {answersTree && answersTree.map(answer => <>
+      <ToCCommentBlock
+        key={answer.item._id}
+        commentTree={answer} indentLevel={1} classes={classes}
+        highlightedCommentId={highlightedLandmarkName}
+      />
+      <Components.TableOfContentsDivider/>
+    </>)}
+    {commentTree && commentTree.map(comment => comment.item
+      ? <ToCCommentBlock
+          key={comment.item._id}
+          commentTree={comment} indentLevel={1} classes={classes}
+          highlightedCommentId={highlightedLandmarkName}
+        />
+      : null
```

Comment:
Are we actually going to have `null`s interleaved here (and if we do, does that cause issues)?

Reply:
Just checked, the type system actually thinks that this null-check isn't necessary.